libs/plugins/a-comp.lv2/a-comp.c

```diff
@@ -665,15 +738,167 @@ render_inline (LV2_Handle instance, uint32_t w, uint32_t max_h)
+}
+
+static void
+render_inline_only_bars (cairo_t* cr, const AComp* self)
+{
+	const float w = self->w;
+	const float h = self->h;
+
+	cairo_rectangle (cr, 0, 0, w, h);
+	cairo_set_source_rgba (cr, .2, .2, .2, 1.0);
+	cairo_fill (cr);
+
+
+	cairo_save (cr);
+
+	const float ht = 0.25f * h;
+
+	const float x1 = w*0.05;
+	const float wd = w - 2.0f*x1;
+
+	const float y1 = 0.17*h;
+	const float y2 = h - y1 - ht;
+
+	cairo_set_source_rgba (cr, 0.5, 0.5, 0.5, 0.5);
+
+	cairo_rectangle (cr, x1, y1, wd, ht);
+	cairo_fill (cr);
 
+	cairo_rectangle (cr, x1, y2, wd, ht);
+	cairo_fill (cr);
+
+	cairo_set_source_rgba (cr, 0.75, 0.0, 0.0, 1.0);
+	const float w_gr = (self->v_gainr > 60.f) ? wd : wd * self->v_gainr * (1.f/60.f);
+	cairo_rectangle (cr, x1+wd-w_gr, y2, w_gr, ht);
+	cairo_fill (cr);
+
+	if (self->v_lvl_out > -60.f) {
+		if (self->v_lvl_out > 10.f) {
+			cairo_set_source_rgba (cr, 0.75, 0.0, 0.0, 1.0);
+		} else if (self->v_lvl_out > 0.f) {
+			cairo_set_source_rgba (cr, 0.66, 0.66, 0.0, 1.0);
+		} else {
+			cairo_set_source_rgba (cr, 0.0, 0.66, 0.0, 1.0);
+		}
+		const float w_g = (self->v_lvl_out > 10.f) ? wd : wd * (60.f+self->v_lvl_out) / 70.f;
+		cairo_rectangle (cr, x1, y1, w_g, ht);
+		cairo_fill (cr);
+	}
+
+	cairo_set_source_rgba (cr, 1.0, 1.0, 1.0, 1.0);
+
+	const float tck = 0.33*ht;
+
+	cairo_set_line_width (cr, .5);
+
+	for (uint32_t d = 1; d < 7; ++d) {
+		const float x = x1 + (d * wd * (10.f / 70.f));
+
+		cairo_move_to (cr, x, y1);
+		cairo_line_to (cr, x, y1+tck);
+
+		cairo_move_to (cr, x, y1+ht);
+		cairo_line_to (cr, x, y1+ht-tck);
+
+		cairo_move_to (cr, x, y2);
+		cairo_line_to (cr, x, y2+tck);
+
+		cairo_move_to (cr, x, y2+ht);
+		cairo_line_to (cr, x, y2+ht-tck);
+	}
+
+	cairo_stroke (cr);
+
+	const float x_0dB = x1 + wd*(60.f/70.f);
+
+	cairo_move_to (cr, x_0dB, y1);
+	cairo_line_to (cr, x_0dB, y1+ht);
+
+	cairo_rectangle (cr, x1, y1, wd, ht);
+	cairo_rectangle (cr, x1, y2, wd, ht);
+	cairo_stroke (cr);
+
+	cairo_set_line_width (cr, 2.0);
+
+	// visualize threshold
+	const float tr = x1 + wd * (60.f+self->v_thresdb) / 70.f;
+	cairo_set_source_rgba (cr, 0.95, 0.95, 0.0, 1.0);
+	cairo_move_to (cr, tr, y1);
+	cairo_line_to (cr, tr, y1+ht);
+	cairo_stroke (cr);
+
+	// visualize ratio
+	const float reduced_0dB = self->v_thresdb * (1.f - 1.f/self->v_ratio);
+	const float rt = x1 + wd * (60.f+reduced_0dB) / 70.f;
+	cairo_set_source_rgba (cr, 0.95, 0.0, 0.0, 1.0);
+	cairo_move_to (cr, rt, y1);
+	cairo_line_to (cr, rt, y1+ht);
+	cairo_stroke (cr);
+}
+
+static void
+aquire_display(AComp* self, uint32_t w, uint32_t h)
+{
+	cairo_surface_t** current = self->display;
+	while (1) {
+		if (!*current) {
+			break;
+		}
+
+		const uint32_t w_ = cairo_image_surface_get_width (*current);
+		const uint32_t h_ = cairo_image_surface_get_height (*current);
+		if (w == w_ && h == h_) {
+			break;
+		}
+		if (*current && current == self->display + DISP_CACHE_NUM) {
+			cairo_surface_destroy (*current);
+			*current = NULL;
+		}
+		++current;
```

Comment:
Actually if you reach `current == self->display + DISP_CACHE_NUM`, you're already after the end.  The pointer is valid from   self->display + 0 ...   self->display + (DISP_CACHE_NUM - 1).

It may be cleaner to rewrite this just with a for loop  `for (idx = 0; idx < DISP_CACHE_SIZE; ++idx)` and use `self->display[idx]`